opencv_ai/splice/tcp.h

```diff
@@ -0,0 +1,100 @@
+#ifndef OPENCV_AI_SPLICE_TCP_H_
+#define OPENCV_AI_SPLICE_TCP_H_
+
+#include <opencv_ai/splice/context.h>
+#include <boost/asio/placeholders.hpp>
+#include <boost/asio/ip/tcp.hpp>
+#include <boost/asio/connect.hpp>
+#include <boost/asio/read.hpp>
+#include <boost/asio/coroutine.hpp>
+#include <boost/bind.hpp>
+
+#include <glog/logging.h>
+
+namespace opencv_ai {
+using tcp = boost::asio::ip::tcp;
+constexpr const size_t kBufferHeaderLength = 4 + 8 + 4;
+
+class BufferClient {
+ public:
+  BufferClient(Context ctx, std::string host, std::string port,
+               std::function<void(size_t stamp,
+                                  const std::vector<char>& buffer)> handler)
+      : ctx_(ctx),
+        host_(host),
+        port_(port),
+        resolver_(ctx_.GetIoService()),
+        socket_(ctx_.GetIoService()),
+        handler_(std::move(handler)) {
+    CHECK(!!handler_) << "Handler may not be null.";
+    tcp::resolver::query query(tcp::v4(), host, port);
+    resolver_.async_resolve(
+        query, ctx_.Wrap(boost::bind(&BufferClient::HandleResolve, this,
+                                     boost::asio::placeholders::error,
+                                     boost::asio::placeholders::iterator)));
+  }
+
+ private:
+  void HandleResolve(const boost::system::error_code& ec,
+                     tcp::resolver::iterator endpoint_iterator) {
+    CHECK(!ec) << "Could not connect to tcp://" << host_ << ":" << port_
```

Comment:
To me this seems in the spirit of crash only software. http://dslab.epfl.ch/pubs/crashonly.pdf 
candea and fox

By CHECK failing here, we are handling the error, and exiting the program... punting to the parent process.

I'm also fine with starting to think about error handling strategies,  but that complicates the class considerably, and was beyond my 30 minute hack session.

If the resolution or the connect fails, then a retry loop with back off might be appropriate. If the recv fails, then reconnect loop might also work.  At some point though, it may just not make sense to retry.

The simplicity of stopping the program immediately, even upon runtime failures is seems so elegant to me, and not sure that the error tolerance buys much in the end.